[GStreamer] Reduce WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebRTC endpoint

philn opened the pull request:
#### ba2d6da1d4c521a644df32594d2ecec23f701597
<pre>
[GStreamer] Reduce WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebRTC endpoint
<a href="https://bugs.webkit.org/show_bug.cgi?id=286270">https://bugs.webkit.org/show_bug.cgi?id=286270</a>

Reviewed by Xabier Rodriguez-Calvar.

The get-transceivers signal and logic to iterate on transceivers was refactored to a new
forEachTransceiver() utility function. If an early break from the loop is desired, the callback
should return true.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::transceiverStatesFromWebRTCBin):
(WebCore::GStreamerMediaEndpoint::doSetLocalDescription):
(WebCore::GStreamerMediaEndpoint::doSetRemoteDescription):
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::pickAvailablePayloadType):
(WebCore::GStreamerMediaEndpoint::transceiverBackendFromSender):
(WebCore::GStreamerMediaEndpoint::collectTransceivers):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::forEachTransceiver):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:

Canonical link: <a href="https://commits.webkit.org/289496@main">https://commits.webkit.org/289496@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fef6db7587b704a209c4ab452c281de3685b541

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87172 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6682 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92033 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/37911 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89222 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/6958 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/14751 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/67374 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/25115 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90174 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5330 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/78901 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47694 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5103 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33279 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37028 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/75586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34156 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/93918 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14334 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/10425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/76174 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14538 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74757 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75376 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18544 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19718 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18158 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7252 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14353 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19646 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14098 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17541 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/15879 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -760,6 +760,31 @@ Seconds StatsTimestampConverter::convertFromMonotonicTime(Seconds value) const
+    }
+
+    // Don't free segments because they're moved to local GRefPtrs.
+    g_array_free(transceivers, FALSE);
```
From what I understand, you're leaking the array segment here. The g_array contains pointers to the GstWebRTCRTPTransceiver objects so you're taking ownership of here is the memory pointed to that pointer. The array contains pointers. You can and actually have to free the array segment.

philn replied:
In the loop above we "adopt" those transceiver refs though, so if we free here wouldn't that be a double free?

philn replied:
If I ask to free the segments:

```
Thread 1 (Thread 0x7f88a9633e00 (LWP 445443)):
#0  _g_log_abort (breakpoint=<optimized out>) at ../glib/gmessages.c:426
#1  g_logv (log_domain=0x7f88ac953af1 "GStreamer", log_level=G_LOG_LEVEL_CRITICAL, format=<optimized out>, args=args@entry=0x7ffc311e2ed0) at ../glib/gmessages.c:1273
#2  0x00007f88ad63e2e3 in g_log (log_domain=log_domain@entry=0x7f88ac953af1 "GStreamer", log_level=log_level@entry=G_LOG_LEVEL_CRITICAL, format=format@entry=0x7f88ac95e418 "\nTrying to dispose object \"%s\", but it still has a parent \"%s\".\nYou need to let the parent manage the object instead of unreffing the object directly.\n") at ../glib/gmessages.c:1315
#3  0x00007f88ac873c90 in gst_object_dispose (object=0x1a861100) at ../gstreamer/subprojects/gstreamer/gst/gstobject.c:408
#4  0x00007f88ad75dc38 in g_object_unref (_object=0x1a861100) at ../gobject/gobject.c:4415
#5  0x00007f88ad5fda7c in array_free (array=0x1a9859f0, flags=FREE_SEGMENT) at ../glib/garray.c:541
#6  0x00007f88b5fa2bcd in WebCore::forEachTransceiver(WTF::GRefPtr<_GstElement> const&, WTF::Function<bool (WTF::GRefPtr<_GstWebRTCRTPTransceiver>&&)>&&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#7  0x00007f88b5f8278f in WebCore::GStreamerMediaEndpoint::requestPad(WTF::GRefPtr<_GstCaps> const&, WTF::String const&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#8  0x00007f88b5f80569 in WebCore::GStreamerMediaEndpoint::linkOutgoingSources(GstSDPMessage*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#9  0x00007f88b5f809b5 in WebCore::GStreamerMediaEndpoint::doSetLocalDescription(WebCore::RTCSessionDescription const*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#10 0x00007f88b5f8b9bd in WebCore::GStreamerPeerConnectionBackend::doSetLocalDescription(WebCore::RTCSessionDescription const*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#11 0x00007f88b5f633d7 in WTF::Detail::CallableWrapper<WebCore::RTCPeerConnection::setLocalDescription(std::optional<WebCore::RTCLocalSessionDescriptionInit>&&, WTF::Ref<WebCore::DeferredPromise, WTF::RawPtrTraits<WebCore::DeferredPromise>, WTF::DefaultRefDerefTraits<WebCore::DeferredPromise> >&&)::$_0, void, WTF::Ref<WebCore::DeferredPromise, WTF::RawPtrTraits<WebCore::DeferredPromise>, WTF::DefaultRefDerefTraits<WebCore::DeferredPromise> >&&>::call(WTF::Ref<WebCore::DeferredPromise, WTF::RawPtrTraits<WebCore::DeferredPromise>, WTF::DefaultRefDerefTraits<WebCore::DeferredPromise> >&&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#12 0x00007f88b5f59beb in WebCore::RTCPeerConnection::chainOperation(WTF::Ref<WebCore::DeferredPromise, WTF::RawPtrTraits<WebCore::DeferredPromise>, WTF::DefaultRefDerefTraits<WebCore::DeferredPromise> >&&, WTF::Function<void (WTF::Ref<WebCore::DeferredPromise, WTF::RawPtrTraits<WebCore::DeferredPromise>, WTF::DefaultRefDerefTraits<WebCore::DeferredPromise> >&&)>&&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#13 0x00007f88b5f59fdd in WebCore::RTCPeerConnection::setLocalDescription(std::optional<WebCore::RTCLocalSessionDescriptionInit>&&, WTF::Ref<WebCore::DeferredPromise, WTF::RawPtrTraits<WebCore::DeferredPromise>, WTF::DefaultRefDerefTraits<WebCore::DeferredPromise> >&&) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#14 0x00007f88b5a3cda1 in WebCore::jsRTCPeerConnectionPrototypeFunction_setLocalDescription(JSC::JSGlobalObject*, JSC::CallFrame*) () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libwebkit2gtk-4.1.so.0
#15 0x00007f884ee0c038 in ??? ()
#16 0x00007ffc311e3410 in ??? ()
#17 0x00007f88b022a397 in llint_op_call_ignore_result () at /var/home/phil/WebKit/WebKitBuild/GTK/Release/lib/libjavascriptcoregtk-4.1.so.0
#18 0x0000000000000000 in ??? ()

STDERR: Trying to dispose object "webrtctransceiver0", but it still has a parent "webkit-webrtcbin-0".
STDERR: You need to let the parent manage the object instead of unreffing the object directly.
```

philn replied:
With the fixup pushed to this PR, the crash no longer absent, and we actually fix an old mem leak as well.

philn committed:
[GStreamer] Reduce WTF_ALLOW_UNSAFE_BUFFER_USAGE in WebRTC endpoint
https://bugs.webkit.org/show_bug.cgi?id=286270

Reviewed by Xabier Rodriguez-Calvar.

The get-transceivers signal and logic to iterate on transceivers was refactored to a new
forEachTransceiver() utility function. If an early break from the loop is desired, the callback
should return true.

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::transceiverStatesFromWebRTCBin):
(WebCore::GStreamerMediaEndpoint::doSetLocalDescription):
(WebCore::GStreamerMediaEndpoint::doSetRemoteDescription):
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::pickAvailablePayloadType):
(WebCore::GStreamerMediaEndpoint::transceiverBackendFromSender):
(WebCore::GStreamerMediaEndpoint::collectTransceivers):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::forEachTransceiver):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:

Canonical link: https://commits.webkit.org/289496@main
